### PR TITLE
Improve docstrings + examples for `as_primitive_array` cast functions

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

N/A
# Rationale for this change
 
I still see people use the more verbose `as_any().downcast_ref::<Type>().unwrap` code; I also used this frequently in my early arrow days, so I though highlighting and improving the docs for the convenience typecasts might help

# What changes are included in this PR?

Add some docstrings and examples

# Are there any user-facing changes?

More docs

### CODE REVIEW

File: arrow/src/array/mod.rs [hunk 1]
```diff
@@ -42,6 +44,19 @@
...
+//! Additionally, there are convenient functions to do this casting
+//! such as [`as_primitive_array<T>`] and [`as_string_array`]:
+//!
+//! ```
+//! # use arrow::array::*;
+//! # use arrow::datatypes::*;
+//! #
+//! fn as_f32_slice(array: &dyn Array) -> &[f32] {
+//!     // use as_primtive_array
+//!     as_primitive_array<Float32Type>().values()
```
Comment (viirya) [inline arrow/src/array/mod.rs:56]: typo?

Comment (viirya) [inline arrow/src/array/mod.rs:56]:
```suggestion
//!     as_primitive_array::<Float32Type>().values()
```

**Review by viirya — APPROVED** [review]

File: arrow/src/array/mod.rs (hunk 1 above)
Comment (alamb) [reply arrow/src/array/mod.rs:56]: Indeed 👍

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Merging as clippy failure unrelated to this PR

**Pull request merged**